### ISSUE TITLE
"Extract Method" crashes in presence of expression declarations (out var, patterns)

### ISSUE THREAD

**Issue by gafter** [dotnet/roslyn#13252]
**Version Used**: 
Roslyn master 2016-08-18

**Steps to Reproduce**:

Make a project containing the program

```cs
using System;

class Program
{
    public static void Main(string[] args)
    {
        int r;
        r = M1(out int y) + M2(3 is int z);
        Console.WriteLine(r);
        Console.WriteLine(y);
        Console.WriteLine(z);
    }
    public static int M1(out int x) { return (x = 3); }
    public static int M2(bool b) { return 4; }
}
```

2. Attempt "extract method" on the line assigning to `r`

**Expected Behavior**:
Good method extracted

**Actual Behavior**:
Exception thrown in IDE, (and if ignored) an incorrect refactoring with duplicate declarations for y and z.

**Comment by jcouv** [comment]
@gafter What do you think the "good method" should be?

The method which is currently (incorrectly) produced:
```C#
    public static void Main(string[] args)
    {
        int r;
        int y, z;
        NewMethod(out r, out y, out z);
        // ...
    }

    private static void NewMethod(out int r, out int y, out int z)
    {
        r = M1(out int y) + M2(3 is int z);
    }
```

The problem with `y` can be fixed by invoking `M1(out y)` instead.
For `z`, I'm not sure a safe refactoring is possible. Any ideas?

@CyrusNajmabadi FYI

**Comment by CyrusNajmabadi** [comment]
If a 'safe' refactoring is not possible, then we should put a warning annotation in the tree so that the preview lets the user know that something is potentially wrong.

**Comment by jcouv** [comment]
@CyrusNajmabadi Do you have an existing example of such a warning? I'm wondering if that is strong enough, or maybe we should block the refactoring in such cases (with is-patterns).

**Comment by CyrusNajmabadi** [comment]
Yes.  We have an annotation called "ConflictAnnotation".  You can see its usage in other refactorings (like inline temporary).  For example, if you had the following code:

```c#
var v = 0;
Foo(ref v);
```

We allow you to inline 'v', but we put the conflict annotation on the usage in 'ref v' since it will produce broken code.

> or maybe we should block the refactoring in such cases (with is-patterns).

Blocking a refactoring is only a 'last resort' hammer to these problems.  The issue is that when a refactoring is blocked the user has no idea why.  They just feel that the IDE is 'broken' and they have no clue what they could even do to fix it.

Also, recall that the IDE goals are different from compiler goals.  You want to error in bad situations.  But we want to be permissive and allow the user to make forward progress toward their end goal.  Even if we can't transform some bits of code, we'd still like to do a lot of hte heavy lifting; allowing hte user to do any final fixup.

**Comment by CyrusNajmabadi** [comment]
Also, to be clear "is patterns" are going to be normal features in the language.  Blocking the refactoring for very esoteric cases is already a hard sell.  Blocking it for bog standard features is not something we want to do at all.

**Comment by jcouv** [comment]
This was fixed by #13874.
